src/async-wrap.cc

```diff
@@ -407,6 +407,40 @@ static void DisablePromiseHook(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+class AsyncWrap::DestroyParam {
+ public:
+  double asyncId;
+  v8::Persistent<Function> callback;
+};
+void AsyncWrap::WeakCallback(
+  const v8::WeakCallbackInfo<AsyncWrap::DestroyParam> &info
+) {
+  AsyncWrap::DestroyParam* p = info.GetParameter();
+  Local<Function> fn = PersistentToLocal(info.GetIsolate(), p->callback);
+  Local<v8::Value> asyncId = v8::Number::New(info.GetIsolate(), p->asyncId);
+  delete p;
+
+  Local<v8::Primitive> undef = Undefined(info.GetIsolate());
+  fn->Call(
+    info.GetIsolate()->GetCurrentContext(), undef, 1, &asyncId);
+}
+static void RegisterDestroyHook(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args[0]->IsObject());
+  CHECK(args[1]->IsNumber());
+  CHECK(args[2]->IsFunction());
+
+  Environment* env = Environment::GetCurrent(args);
+
+  v8::Persistent<Object> target(env->isolate(), args[0].As<Object>());
```

Comment:
I think this might be causing trouble because there’s no way to reset the persistent if it’s stack-allocated? Maybe that’s not an issue, but I’d prefer it if this was a property on `DestroyParam`

Reply:
We're only using this to register the weak callback and let it get cleaned up afterwards. Do we need to keep it alive for v8?

Reply:
It’s making me a bit uneasy at least in the sense that there are comments in v8.h that hint at a future change towards `v8::Persistent` resetting itself when the destructor is called (which I would welcome, I think):

https://github.com/nodejs/node/blob/b204b09dae7df56f03252c70fd28ba42d09e5197/deps/v8/include/v8.h#L633-L634